grove/alpha/simon/utils.py

```diff
@@ -0,0 +1,70 @@
+import numpy as np
+
+PADDED_BINARY_BIT_STRING = "{0:0{1:0d}b}"
+
+
+def is_unitary(matrix):
+    """
+    A helper function that checks if a matrix is unitary.
+
+    :param 2darray matrix: a matrix to test unitarity of
+    :return: true if and only if matrix is unitary
+    :rtype: bool
+    """
+    rows, cols = matrix.shape
+    if rows != cols:
+        return False
+    return np.allclose(np.eye(rows), matrix.dot(matrix.T.conj()))
+
+
+def most_significant_bit(lst):
+    """
+    A helper function that finds the position of
+    the most significant bit in a 1darray of 1s and 0s,
+    i.e. the first position where a 1 appears, reading left to right.
+
+    :param 1darray lst: a 1darray of 0s and 1s with at least one 1
+    :return: the first position in lst that a 1 appears
+    :rtype: int
+    """
+    return np.argwhere(np.asarray(lst) == 1)[0][0]
+
+
+def bitwise_xor(bs0, bs1):
+    """
+    A helper to calculate the bitwise XOR of two bit string
+
+    :param bs0: String of 0's and 1's representing a number in binary representations
+    :param bs1: String of 0's and 1's representing a number in binary representations
+    :return: String of 0's and 1's representing the XOR between bs0 and bs1
+    :rtype: str
+    """
+    assert len(bs0) == len(bs1), "Bit strings are not of equal length"
+    n_bits = len(bs0)
+    return PADDED_BINARY_BIT_STRING.format(int(bs0, 2) ^ int(bs1, 2), n_bits)
+
+
+def binary_back_substitute(W, s):
```

Comment:
Variable names should be more descriptive, and not be capitalized.

Reply:
While I agree with you in general, I'd argue that in this case it is warranted. See, e.g. the scikit-learn API. The reason for this is that in this way it represents the mathematical structure of the underlying problem that is being solved, which is a matrix equation. In this context a matrix is almost always capitalized and a vector is lower-case. This also makes it easier to match with the docstrings. In light of clarifying code intention I'd rather overrule the more general PEP8 for specific highlighting.